packages/firestore/src/model/proto_values.ts

```diff
@@ -64,10 +64,10 @@ export function typeOrder(value: api.Value): TypeOrder {
 
 /** Returns whether `value` is defined and corresponds to the given type order. */
 export function isType(
-  value: api.Value | undefined,
+  value: api.Value | null | undefined,
   expectedTypeOrder: TypeOrder
-): boolean {
-  return value !== undefined && typeOrder(value) === expectedTypeOrder;
+): value is api.Value {
```

Comment:
It tells the transpiler that it can assume that `value is api.Value ` if the function returns true. This allows the following:

```
let valueOrUnknown : api.Value | null | undefined = ...;

if (isType(valueOrUnknown, ...)) {
   // valueOrUnknown is api.Value
}
```

This is all based on the honor system and we could easily fool the transpiler by providing the wrong guarantees.